src/uct/tcp/tcp_iface.c

```diff
@@ -218,6 +217,23 @@ static int uct_tcp_iface_is_reachable(const uct_iface_h tl_iface,
     return 1;
 }
 
+static int
+uct_tcp_iface_is_reachable_v2(const uct_iface_h iface,
+                              const uct_iface_is_reachable_params_t *params)
+{
+    if (!uct_iface_is_reachable_params_valid(
+                params, UCT_IFACE_IS_REACHABLE_FIELD_DEVICE_ADDR)) {
+        return 0;
+    }
+
+    if (!uct_tcp_iface_is_reachable(iface, params->device_addr)) {
+        return 0;
+    }
+
+    /* Scope needs to be checked after interface reachability */
+    return uct_iface_is_reachable_scope(iface, params);
```

Comment:
TCP needs to check `uct_tcp_iface_is_reachable()` before checking for scope with `uct_iface_is_reachable_scope()` as device_addr sizes might not match (found with valgrind).

Do other transports need to use similar order? I yes we could just call a base function from all tranports like:
`return uct_base_iface_is_reachable_v2(iface, params, flags, tl_is_reach_func)`

Reply:
added decorator pattern to make sure to check scope after transport specific reachability checks.